src/unused_finder/mod.rs

```diff
@@ -110,103 +122,269 @@ pub fn find_unused_items(
-                                            unused_files.remove(imported_path);
-                                        }
-                                        _ => {
-                                            unused_files.remove(imported_path);
-                                            let i = ExportedItem::from(imported);
-                                            origin_file_exported_items.remove(&i);
-                                        }
+                                ImportedItem::ExecutionOnly | ImportedItem::Namespace => {
+                                    removed.import_export_info.exported_ids.clear();
+                                }
+                                _ => {
+                                    removed
+                                        .import_export_info
+                                        .exported_ids
+                                        .remove(&node_visitor::ExportedItem::from(&item));
+                                }
+                            }
+                        }
+                        used_files.insert(p, removed);
+                    }
+                    None => {
+                        if let Some(used_file) = used_files.get_mut(&p) {
+                            for item in items {
+                                match item {
+                                    ImportedItem::ExecutionOnly | ImportedItem::Namespace => {
+                                        used_file.import_export_info.exported_ids.clear();
+                                    }
+                                    _ => {
+                                        used_file
+                                            .import_export_info
+                                            .exported_ids
+                                            .remove(&node_visitor::ExportedItem::from(&item));
                                     }
                                 }
-                                ResolvedItem::Exported(_) => {}
                             }
-                        });
+                        }
                     }
-                });
-        });
-    let mut unused_items_len = 0;
-
-    // Print the unused items for each file, sort them by file first
-    let mut path_unused_items: Vec<_> = path_unused_items_map.into_iter().collect();
-    path_unused_items.sort_by(|x, y| x.0.cmp(&y.0));
-    path_unused_items.iter().for_each(|(k, v)| {
-        if v.len() > 0 {
-            if !unused_files.contains(k) {
-                println!("From file {} found {:?} unused items", k, v);
+                }
             }
-        }
-        unused_items_len += v.len();
+        });
     });
 
+    loop {
+        let old_size = unused_files.len();
+        let mut new_used_files: HashMap<String, &mut WalkFileMetaData> = HashMap::new();
+
+        used_files.iter_mut().for_each(|(_used_file_path, file)| {
```

Comment:
This looks like, essentially, a frontier walk of a directed graph using weird terms
With that observation:

- It looks like we re-traverse `used_files` every iteration, even though we have already traversed _most_ of `used_files`. What's the reason? Am I missing something here?
- It would be faster to maintain an explicit `new_used_files` (frontier set) and loop through that instead (while continuing to build `used_files` as the visited set to avoid cycles)

Reply:
> It looks like we re-traverse used_files every iteration, even though we have already traversed most of used_files. What's the reason? Am I missing something here?
Just did that because of some problems I had with the borrow checker, will fix that in a later push.